src/lib/storage/buffer/volatile_region.hpp

```diff
@@ -0,0 +1,105 @@
+ */
+namespace hyrise {
+class VolatileRegion final : public Noncopyable {
+ public:
+  // Default size of the virtual memory region for all page size types. Default is 256 GB.
+  constexpr static uint64_t DEFAULT_RESERVED_VIRTUAL_MEMORY = 1UL << 38;
+  constexpr static uint64_t DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION =
+      (DEFAULT_RESERVED_VIRTUAL_MEMORY / PAGE_SIZE_TYPES_COUNT) / bytes_for_size_type(MAX_PAGE_SIZE_TYPE) *
+      bytes_for_size_type(MAX_PAGE_SIZE_TYPE);
+
+  using MappedRegion = std::span<std::byte, DEFAULT_RESERVED_VIRTUAL_MEMORY>;
+  using MappedPageRegion = std::span<std::byte, DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION>;
+
+  // Enable mprotect class for debugging purposes
+  constexpr static bool ENABLE_MPROTECT = false;
+
+  // Create a VolatileRegion in a memory region for a given size_type.
+  VolatileRegion(const PageSizeType size_type, MappedPageRegion region);
+
+  // Get the frame of a given page.
+  Frame* get_frame(const PageID page_id);
+
+  // Get the start address of a given page.
+  std::byte* get_page(const PageID page_id);
+
+  // Use mbind for page movement. This is not portable and only works on Linux with NUMA support.
+  void mbind_to_numa_node(const PageID page_id, const NodeID target_memory_node);
+
+  // Use move_pages for page movement. This is not portable and only works on Linux with NUMA support.
+  void move_page_to_numa_node(const PageID page_id, const NodeID target_memory_node);
+
+  // Use double memcpy + madvise(MADV_DONTNEED) with intermediate buffer for page movement. This is faster with more concurrency.
+  void memcpy_page_to_numa_node(const PageID page_id, const NodeID target_memory_node);
+
+  // Free a page using madvise(MAV_FREE_REUSABLE) on OS X and madvise(MADV_DONTNEED) on Linux.
+  void free(const PageID page_id);
+
+  // Mark a page as reusable using madvise(MADV_FREE_REUSE) on OS X to update memory accounting. Not implemented on Linux.
+  void reuse(const PageID page_id);
+
+  // Returns the number of pages this region can manage.
+  size_t size() const;
+
+  // Returns the size type of this region.
+  PageSizeType size_type() const;
+
+  // Calculate the approximate memory used by this object.
+  size_t memory_consumption() const;
+
+  // Create a fixed-sized memory region using mmap that can be divided into regions of pages for all size types.
+  static MappedRegion create_mapped_region();
+
+  // Unmap the fixed-sized memory region
+  static void unmap_region(MappedRegion region);
+
+  // Create a VolatileRegion for each size type from the fixed-sized memory region.
+  static std::array<std::shared_ptr<VolatileRegion>, PAGE_SIZE_TYPES_COUNT> create_volatile_regions(
```

Comment:
I think a little visualization or at least explanation in the header of what the connection between VolatileRegion, MappedRegion, and MappedPageRegion is, would be extremely helpful.